infrastructure/README.md

```diff
@@ -18,75 +18,184 @@ Before running the infra scripts, you need to:
+    * `kube_context`: Local kuberenetes context containing authentication to the target cluster
 
    
-4. You could optionally set the following variables if required 
+   You could optionally set the following variables if required
     * `aws_profile`: Set this if you are not using the default account set with AWS CLI
     * `quay_image`: Overrides the image that is being used 
+    * `dns_domain`: Alternative domain name to be created along with the default (default is `<prefix>.quaydev.org`)
     * `clair_image`: Overrides the image that being used
-    * `quay_vpc_cidr`: CIDR of VPC where quay resources like DB, redis will be deployed
     * `builder_ssh_keypair`: SSH Keypair created to access the build VMs (should be created prior to deploy)
     * `builder_access_key`: AWS access key for builder. Used to createEC2 VMs for building
     * `builder_secret_key`: AWS Secret key for builder. Used to createEC2 VMs for building
+    * `redis_azs`: List of Redis availability zones
+
+1. If using an env file like the examples given, set the environment variables in the current shell `source envs/example-primary.env`.
+    > **NOTE** The example env file sets the `kube_context` with the command `oc config current-context` so the OCP cluster needs to be logged into first before sourcing the environment file.
+
+1. Initialize Terraform `terraform init`
+
+1. Create the resources `terraform apply`
+    > **NOTE** Terraform also generates a statefile `terraform.tfstate`. DO NOT DELETE this file or commit it. This file keeps track of all the resources on AWS associated with your workspace.
 
-The easiest way to get started is to use the provided environment variable samples in the `envs` directory
+1. This command generates all the resources required and outputs `<prefix>_quay_deployment.yaml` file which you can deploy to openshift.
 
-## Running
+    ```
+    oc apply -f <prefix>-quay-deployment.yaml
+    ```
 
-The following gives an example of creating a new environment from scratch
+    This will generate all the deployments for Quay.
 
-```bash
-$ terraform workspace new syed-py3
-$ terraform workspace select syed-py3
-$ oc login <login token>
-$ source envs/syed-py3.env
-$ terraform apply
+
+1. Once you get quay running, you can get the quay endpoint by running
+
+    ```
+    $ oc project <prefix>-quay
+    $ oc get route
+    ```
+
+## Creating the secondary instance
+
+**NOTE** Creating the secondary instance requires completing the setup of the primary instance from the previous section first.
+
+1. The secondary instance is readonly and requires service keys to be configured through the primary instance. The service keys are used for functions related to signing and token generation. First we need to generate those keys via a script in the Quay repository.
+
+```
+$ git clone https://github.com/quay/quay  # clone the Quay repository
```

Comment:
you can generate this in terraform. Just curl that file in a `null` resource and add the keys to the template

Reply:
Right now there's no way to output read information from the null resource. I tried outputting to a file and reading from that but the file command expects the file to exist before terraform apply is ran. There's also the external resource but that runs on the plan and before the apply and since this runs commands against the primary DB I opted to not do that.

I instead added the service key setup to a script that can be ran and then pass the kid and pem as parameters which will get templated, to make things easier. 

There might be a way to get the parameter and reference it in the template generation but Terraform doesn't do well with running scripts like this imo.

Reply:
Here's my take
```
locals {
  prefix = "testing"
}

data "http" "genkeyscript" {
  url = "https://raw.githubusercontent.com/quay/quay/master/tools/generatekeypair.py"
}

resource "null_resource" "run_python_script" {
  provisioner "local-exec" {
    command = "python -c '${data.http.genkeyscript.body}' '${local.prefix}'"
  }
}

data "local_file" "key" {
  filename = "${local.prefix}.jwk"
}

data "local_file" "cert" {
  filename = "${local.prefix}.pem"
}

data "local_file" "key_id" {
  filename = "${local.prefix}.kid"
}

output "key" {
  value = data.local_file.key.content
}

output "cert" {
  value = data.local_file.cert.content
}

output "key_id" {
  value = data.local_file.key_id.content
}

resource "null_resource" "cleanup" {
  provisioner "local-exec" {
    command = "rm -f '${local.prefix}.jwk' '${local.prefix}.pem' '${local.prefix}.kid'"
  }
  depends_on = [data.local_file.key, data.local_file.cert, data.local_file.key_id]
}
```
save this in a file `main.tf` and run `terraform apply`

Reply:
Yup got it to work with the `data.local_file` resource, updated.